src/reanimated2/component/KeyboardAvoidingView.tsx

```diff
@@ -0,0 +1,165 @@
+import React, { forwardRef, useCallback } from 'react';
+import type {
+  KeyboardAvoidingViewProps,
+  LayoutRectangle,
+  View,
+  ViewProps,
+} from 'react-native';
+import { StyleSheet, useWindowDimensions } from 'react-native';
+import { AnimatedView as ReanimatedView } from './View';
+import {
+  useAnimatedKeyboard,
+  useAnimatedStyle,
+  runOnUI,
+  useSharedValue,
+} from '../';
+
+let useHeaderHeight: () => number;
+try {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  useHeaderHeight = require('@react-navigation/elements').useHeaderHeight;
+} catch (e) {
+  useHeaderHeight = () => 0;
+}
+
+type Props = ViewProps & KeyboardAvoidingViewProps;
+
+/**
+ * View that moves out of the way when the keyboard appears by automatically
+ * adjusting its height, position, or bottom padding.
+ */
+const KeyboardAvoidingView = forwardRef<View, React.PropsWithChildren<Props>>(
+  (
+    {
+      behavior,
+      children,
+      contentContainerStyle,
+      enabled = true,
+      keyboardVerticalOffset = 0,
+      style,
+      onLayout,
+      ...props
+    },
+    ref
+  ) => {
+    const initialFrame = useSharedValue<LayoutRectangle | null>(null);
+    const keyboard = useAnimatedKeyboard();
+    const { height: screenHeight } = useWindowDimensions();
+    const headerHeight = useHeaderHeight();
+
+    const onLayoutWorklet = useCallback(
+      (layout: LayoutRectangle) => {
+        'worklet';
+        if (initialFrame.value == null) {
+          initialFrame.value = layout;
+        }
+      },
+      [initialFrame]
+    );
+
+    const handleOnLayout = useCallback<NonNullable<ViewProps['onLayout']>>(
+      (event) => {
+        runOnUI(onLayoutWorklet)(event.nativeEvent.layout);
+
+        if (onLayout) {
+          onLayout(event);
+        }
+      },
+      [onLayout, onLayoutWorklet]
+    );
```

Comment:
Do we maybe want the event to go through Reanimated's event emitter? cc @tomekzaw

Reply:
`runOnUI` doesn't look right here indeed